pageserver/src/tenant/layer_map/historic_layer_coverage.rs

```diff
@@ -415,6 +425,55 @@ impl<Value: Clone> BufferedHistoricLayerCoverage<Value> {
         self.buffer.insert(layer_key, None);
     }
 
+    /// Replaces a previous layer with a new layer value.
+    ///
+    /// The replacement is conditional on:
+    /// - there is an existing `LayerKey` record
+    /// - there is no buffered removal for the given `LayerKey`
+    /// - the given closure returns true for the current `Value`
```

Comment:
These different conditions, to me, anyways, are another argument for having an enum as a result type instead of a bool.
Slap a `Debug` on that enum, and the call site in `timeline.rs` can print it in the `error!()`.
Better readability and much easier debugging if the replacement ever fails in prod.

Reply:
```rust
enum Replacement<Value> {
  /// Old value was replaced with the given new value.
  Replaced {
    /// `true` if the key had a
    in_buffered: boolean
  },

  /// Key was not found in buffered or existing layers.
  NotFound,

  /// A removal for the replaced key has been buffered so no replacement was done.
  RemovalBuffered,

  /// Value was rejected by the given closure.
  Unexpected(Value),
}
```

Reply:
[38c861b](https://github.com/neondatabase/neon/pull/3513/commits/38c861b40a7fcbce10b49e5b9190bc8036964b2e)

Reply:
Interesting borrowck failure, rust-analyzer doesn't show it (had began to rely on it).

Reply:
Ok it's nll/polonius issue 1.

Reply:
I'll just clone the arc.